### PR TITLE
Implements Custom Parameters in Data Unit

### LINKED ISSUE #12: `DecodeStream` lacks support for many `custom_x_flag` sequence header fields
Many of these overrides are not supported (despite generally being purely
metadata and not requiring any special decoder behaviour). The following test
bitstreams exercise all permissible settings of these flags.

[custom_flags.zip](https://github.com/bbc/vc2-reference/files/4823996/custom_flags.zip)

```
$ DecodeStream custom_flags_1.vc2 out.16p2
$ DecodeStream custom_flags_2.vc2 out.16p2
Error: DataUnitIO: custom_pixel_aspect_ratio_flag set, shouldn't be
$ DecodeStream custom_flags_3.vc2 out.16p2
Error: DataUnitIO: Invalid Frame Rate on Input: 0
$ DecodeStream custom_flags_4.vc2 out.16p2
Error: DataUnitIO: Invalid Frame Rate on Input: 0
```

NB: The first test file sets all custom flags to false (and is correctly decoded by the decoder). The remaining test files encode exactly the same video format but set some or all custom value flags.

### LINKED ISSUE #15: `Encode*` does not reflect top or bottom field first settings (`-t` and `-b`) in bitstream
When encoding a bitstream, the top-field-first flag is determined by the base
video format selected. The various `Encode*` implementations select the base
video format based on unrelated criteria.

In the examples below the following [sample frame (1 frame, 10 bit, 1280x720, 4:2:2, progressive, YCbCr)](https://github.com/bbc/vc2-reference/files/4823981/linear_ramps.zip)
is used.

Below the `-t` argument is given (indicating that top-field-first should be
selected) however the encoder chooses base video format 0 (custom format) which
defines top field first to be false. 

    EncodeHQ-CBR \
        -t \
        -x 1280 -y 720 -f 4:2:2 -l 10 -r 1 \
        -k LeGall -d 3 -a 2 -u 1 -s $((1280*720*2*2/4)) -S 2 \
        input.16p2 /tmp/out.vc2

The custom base format is chosen by the encoder in this case because the
resolution/subsampling/framerate combination does not correspond with an
existing base video format. (As an aside, the most efficient encoding would be
to pick base video format 10 and override the frame rate -- this would also set
the field order correctly in this case.)

As an example of the opposite problem, below the `-b` argument is given (bottom
field first) but the encoder chooses base video format 10 (HD 720p 50) which
sets top-field-first to true.

    EncodeHQ-CBR \
        -b \
        -x 1280 -y 720 -f 4:2:2 -l 10 -r 6 \
        -k LeGall -d 3 -a 2 -u 1 -s $((1280*720*2*2/4)) -S 2 \
        input.16p2 /tmp/out.vc2

This time the video format specified matches base format 10 and hence the
encoder picking that format, though the field order specification is ignored.

### PR DESCRIPTION
**Pull request by GalenReich**
Closes #15 

This pull request causes top and bottom field first settings to be respected in the encoders. To have the correct behaviour, the selection of base video format was rewritten: the appropriate level and corresponding base format are selected as before (if appropriate), but when the level is zero, the base video format is chosen to require the fewest custom flags to achieve the desired custom format. As there is no override for the top-field-first, the selection of base video format is subject to the hard condition that the desired top-field-first parameter matches that of the base format.

Closes #12 

While rewriting the base video format selection code, it made sense to include the other custom parameters from the spec. This pull request therefore also includes the "plumbing" of the custom parameters within the DataUnit file, which results in the decoder being able to handle vc2 streams with a variety of custom parameters. The decoder doesn't implement any special additional behaviour to handle the custom parameters.

### CODE REVIEW

**Comment by GalenReich** [comment]
I had thought about removing the level selection code in DataUnit.cpp, but am trying not to remove too much functionality where possible. If we think that it's not needed (as the reference can produce conformant streams that are level 0) I'm happy to remove it. Otherwise, I'll open an issue to check it for correctness and refactor it.

File: src/EncodeHQ-CBR/EncodeParams.cpp [hunk 1]
```diff
@@ -126,8 +126,8 @@ ProgramParams getCommandLineParams(int argc, char * argv[], const char * details
     const int fragment_length = cla_fragmentLength.getValue();
 
     // Check for valid combinations of parameters and options
-    if ((chromaFormat==RGB) && (cla_lumaDepth.isSet() || cla_chromaDepth.isSet()))
-      throw invalid_argument("luma/chroma depth is not appropriate for RGB (use -z or --bitDepth)");
+    // if ((chromaFormat==RGB) && (cla_lumaDepth.isSet() || cla_chromaDepth.isSet()))
+    //   throw invalid_argument("luma/chroma depth is not appropriate for RGB (use -z or --bitDepth)");
```
Comment (jamesba) [inline src/EncodeHQ-CBR/EncodeParams.cpp:130]: Either remove or do not. Don't leave code commented out in a PR for merging.

File: src/EncodeHQ-ConstQ/EncodeParams.cpp [hunk 2]
```diff
@@ -123,8 +123,8 @@ ProgramParams getCommandLineParams(int argc, char * argv[], const char * details
     const int slicePrefix = cla_slicePrefix.getValue();
 
     // Check for valid combinations of parameters and options
-    if ((chromaFormat==RGB) && (cla_lumaDepth.isSet() || cla_chromaDepth.isSet()))
-      throw invalid_argument("luma/chroma depth is not appropriate for RGB (use -z or --bitDepth)");
+    // if ((chromaFormat==RGB) && (cla_lumaDepth.isSet() || cla_chromaDepth.isSet()))
+    //   throw invalid_argument("luma/chroma depth is not appropriate for RGB (use -z or --bitDepth)");
```
Comment (jamesba) [inline src/EncodeHQ-ConstQ/EncodeParams.cpp:127]: ditto

File: src/EncodeLD/EncodeParams.cpp [hunk 3]
```diff
@@ -122,8 +122,8 @@ ProgramParams getCommandLineParams(int argc, char * argv[], const char * details
     const int fragment_length = cla_fragmentLength.getValue();
 
     // Check for valid combinations of parameters and options
-    if ((chromaFormat==RGB) && (cla_lumaDepth.isSet() || cla_chromaDepth.isSet()))
-      throw invalid_argument("luma/chroma depth is not appropriate for RGB (use -z or --bitDepth)");
+    // if ((chromaFormat==RGB) && (cla_lumaDepth.isSet() || cla_chromaDepth.isSet()))
+    //   throw invalid_argument("luma/chroma depth is not appropriate for RGB (use -z or --bitDepth)");
```
Comment (jamesba) [inline src/EncodeLD/EncodeParams.cpp:126]: And again

File: src/Library/DataUnit.h [hunk 4]
```diff
@@ -94,14 +94,40 @@ class WrappedPicture {
...
+                  int leftOffset = -1,
+                  int topOffset = -1,
+                  
+                  ColorSpec colorSpec = CS_UNSET,
+                  int colorPrimaries = 0, // HDTV
+                  int colorMatrix = 0, // HDTV
+                  int transferFunction = 0, //TV Gamma
+                  
+                  bool use_v3 = false
+                );
```
Comment (jamesba) [inline src/Library/DataUnit.h:130]: Much nicer!

File: src/EncodeHQ-CBR/EncodeParams.cpp (hunk 1 above)
Comment (GalenReich) [reply src/EncodeHQ-CBR/EncodeParams.cpp]: All gone!

**Pull request merged**